[PORT/Revival] Custom say emotes and Emphasis

Bobbanz1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Ports this PR from NSV13
[#2170](https://github.com/BeeStation/NSV13/pull/2170)
However that PR is a port from here so it's technically a revival of #5897 rather than a port?

Ports from Whitesands
[#496](https://github.com/Whitesands13/Whitesands/pull/496)
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Allows for better Roleplaying over the Radio and in person with other characters.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

Will have to add this later but for now have this image from NSV13 of it working

![200263401-e19f782b-3f3b-49b4-8d9e-9a73d63d588b](https://user-images.githubusercontent.com/59128051/204469984-cf9a259b-8ba8-4087-9552-16fc939d75db.png)


</details>

## Changelog
:cl:Kapu1178, Kevinz, XSlayer300, Ryll/Shaps, Seris02, GoldenAlpharex, Shadowtail117
add: Speech Format Help verb under the OOC tab.
add: Text emphasis. Surrounding a block of text with underscores will italicize it and surrounding it with two asterisks will bold it
add: Emotes that start with an apostrophe wont format weirdly
add: You can now have custom say mods, and can use them to emote over radio. say";screeches-HELP SECURITY!" results in [Common] Joe screeches, "HELP SECURITY!"
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Bobbanz1 committed:
Necromancy

Bobbanz1 committed:
Merge branch 'BeeStation:master' into LoudWordsAndThee

PowerfulBacon commented on code/game/say.dm:
```diff
@@ -111,24 +118,38 @@ GLOBAL_LIST_INIT(freqtospan, list(
+/atom/proc/say_emphasis(input)
+	ENCODE_HTML_EMPHASIS(input, "\\|", "i", italics)
+	ENCODE_HTML_EMPHASIS(input, "\\+", "b", bold)
+	ENCODE_HTML_EMPHASIS(input, "_", "u", underline)
```
I hate the + and |, bold should be * I think or the same as markdown \*\***bold**\*\*.

If the problem is that math may be picked up like "8 * 4 + 5 * 2", then we can just do what discord does and only allow it if there is no spaces between it.
```regex
\*\*[^ ](.*)[^ ]\*\*
```

I don't mind having both for people who play on multiple servers, but I think \* should either be bold, or use markdown for italics and boldness.

PowerfulBacon commented on code/game/say.dm:
```diff
@@ -111,24 +118,38 @@ GLOBAL_LIST_INIT(freqtospan, list(
+/atom/proc/say_emphasis(input)
+	ENCODE_HTML_EMPHASIS(input, "\\|", "i", italics)
+	ENCODE_HTML_EMPHASIS(input, "\\+", "b", bold)
+	ENCODE_HTML_EMPHASIS(input, "_", "u", underline)
```
Also I don't think we should allow people to talk in underlines, since bold provides the emphasis that underlined would need and it kind of ends up looking more like a hyperlink than some important text.
![image](https://user-images.githubusercontent.com/26465327/204843996-b8cf0400-31d0-44cb-8938-8006a2d4a91e.png)
If other maintainers disagree with me (I've asked in dev channel), then I'll resolve this.

Bobbanz1 replied:
Underline has been removed

Archanial reviewed (dismissed):
Remove underline entirely
I'll review code in a bit

Bobbanz1 committed:
Right thats that for now

Bobbanz1 committed:
Merge branch 'master' into LoudWordsAndThee

Bobbanz1 committed:
Merge branch 'master' into LoudWordsAndThee

Bobbanz1 committed:
Merge branch 'LoudWordsAndThee' of https://github.com/Bobbanz1/BeeStation-Hornet into LoudWordsAndThee

Bobbanz1 committed:
Revert "Merge branch 'master' into LoudWordsAndThee"

This reverts commit 0ef0737aa9cb94ae40da36543dcf490bef4734a1.

Bobbanz1 committed:
Revert "Revert "Merge branch 'master' into LoudWordsAndThee""

This reverts commit 4304d29fb5c3d3a34df530b631dc52364784d3f0.

Bobbanz1 committed:
Can we go five seconds without accidentally causing havoc?

itsmeow commented on code/datums/emotes.dm:
```diff
@@ -201,3 +203,16 @@
+
+/**
+ * Returns a boolean based on whether or not the string contains a comma or an apostrophe,
+ * to be used for emotes to decide whether or not to have a space between the name of the user
```
I'm really confused in what case you would NOT want a space

Bobbanz1 replied:
Removed the Proc and made it the way it was previously

Kapu1178 replied:
*me's eyeballs FUCKING EXPLODE gets spit out as `Kapu1178 's eyeballs FUCKING EXPLODE` without this.

Bobbanz1 replied:
Ooookay time to reinstall it then!

itsmeow replied:
I just want someone to explain

itsmeow replied:
Oh kapu explained. i thought it was a meme

itsmeow commented on code/game/say.dm:
```diff
@@ -111,24 +118,29 @@ GLOBAL_LIST_INIT(freqtospan, list(
+	return "[say_mod], \"[spanned]\""
+
+/atom/proc/say_emphasis(message, var/list/ignore = list())
+	var/regex/markup
```
can you document what the hell this proc does, because I do not understand by looking at it

Bobbanz1 replied:
Done to the best of my current knowledge of it

itsmeow commented on code/modules/mob/say.dm:
```diff
@@ -101,6 +101,19 @@
 ///The amount of items we are looking for in the message
 #define MESSAGE_MODS_LENGTH 6
 
+/mob/proc/check_for_custom_say_emote(message, list/mods)
```
add a comment for what this does as well, it's a bit confusing

Bobbanz1 replied:
Done to the best of my current knowledge of it, might not be a useful comment but I tried my best.

Bobbanz1 committed:
Updates in response to review!

Also fixed that annoying bug that made messages sent as "screams- Help" appear as screams " Help"

Bobbanz1 committed:
Update ooc.dm

itsmeow reviewed (dismissed):
Good improvement!

Bobbanz1 committed:
Return of the should_have_space_before_emote!

PowerfulBacon reviewed (dismissed):
Briefly skimmed through the code, I don't see anything majorly wrong.
Very good changes to the control characters. I am satisfied.

Archanial commented on code/__DEFINES/say.dm:
```diff
@@ -41,7 +41,9 @@
 #define MODE_KEY_VOCALCORDS "x"
 
-#define MODE_SING "%"
+#define MODE_SING "sing"
```
why is this changed? Does that mean you can't say "sing" without singing?

Kapu1178 replied:
we went over this when i made the original pr. mode != key. its a code only thing to make its purpose clear.

Archanial replied:
it might come as a surprise but I do not remember every answer to my every comment ever

Archanial commented on code/datums/emotes.dm:
```diff
@@ -87,11 +87,13 @@
 
 	user.log_message(msg, LOG_EMOTE)
 
+	var/space = should_have_space_before_emote(html_decode(msg)[1]) ? " " : ""
```
```suggestion
	var/space = should_have_space_before_emote(html_decode(msg)[1]) ? " " : null
```

Kapu1178 replied:
why? typefucking for no reason makes bad code.

Archanial replied:
because no separator (space) should be no separator not an empty string.
Everything in dm is nullable so I don't get the "typefucking" comment.

Archanial commented on code/datums/emotes.dm:
```diff
@@ -201,3 +203,16 @@
+ */
+/proc/should_have_space_before_emote(string)
+	var/static/regex/no_spacing_emote_characters = regex(@"(,|')")
+	return no_spacing_emote_characters.Find(string) ? FALSE : TRUE
```
```suggestion
	return !no_spacing_emote_characters.Find(string)
```

Archanial commented on code/modules/mob/mob.dm:
```diff
@@ -181,7 +181,7 @@
 
 
-/atom/proc/visible_message(message, self_message, blind_message, vision_distance = DEFAULT_MESSAGE_RANGE, list/ignored_mobs, list/visible_message_flags, allow_inside_usr = FALSE)
+/atom/proc/visible_message(message, self_message, blind_message, vision_distance = DEFAULT_MESSAGE_RANGE, list/ignored_mobs, list/visible_message_flags, allow_inside_usr = FALSE, separation = " ")
```
I am unsure if the default should be space. Doesn't this impact other cases than emotes etc?
Also this proc should handle null separator

PowerfulBacon replied:
from the looks of the code, the seperation is just the character that comes between the name and the message. The default was space before, so there should be no behavioural changes from what we have now that isn't intended to be changed by this PR.
The only situation where you would want this as something like null, would be '*me 's suit sensors flicker out.'

Archanial commented on code/modules/mob/say.dm:
```diff
@@ -101,6 +101,34 @@
+	message = copytext(message, customsaypos + messagetextpos)
+	if(!message)
+		mods[MODE_CUSTOM_SAY_ERASE_INPUT] = TRUE
+		message = "an interesting thing to say"
```
Are you sure it's a good idea to force people to say this?
Can you show any screenshots of how this looks?

Kapu1178 replied:
its a debug tool. if you see this the yame is broken.

EvilDragonfiend replied:
Yeah I noticed this is to track when someone says this, but I think you need to put a stacktrace when they say this.

Archanial replied:
Stacktrace is better and will help us fix any bugs later

PowerfulBacon replied:
I agree that this should either CRASH if we are expecting a message to be passed in (and then !message has to be checked every time this proc is called) or we expect message to be nullable and just return empty string or nullable and accept the input.

Bobbanz1 committed:
Apply suggestions from code review

Co-authored-by: Archanial <37456678+Archanial@users.noreply.github.com>

Bobbanz1 committed:
Stack trace introduced as per request of review
